tools/codegen/api/translate.py

```diff
@@ -0,0 +1,143 @@
+#     g({exprs}); // g is a different API
+#   }
+#
+# and you need to generate "exprs".
+#
+# TODO: Don't need full Binding for goals, CType will do
+# TODO: Don't need full Binding for bindings, list of Expr will do
+def translate(bindings: Sequence[Binding], goals: Sequence[Binding], *, method: bool = False) -> List[Expr]:
+    # Add all the bindings to the context
+    ctx: Dict[CType, str] = {}
+    for b in bindings:
+        ctx[b.ctype] = b.name
+
+    # Add implicit bindings if the generated code is inside a Tensor method
+    if method:
+        ctx[MutRefCType(BaseCType("Tensor", "self"))] = "const_cast<Tensor&>(*this)"
+        ctx[ConstRefCType(BaseCType("Tensor", "self"))] = "const_cast<Tensor&>(*this)"
+        # This is better!  Byte-for-byte compat
+        # ctx[ConstRefCType(BaseCType("Tensor", "self"))] = "*this"
+
+    def unsat(goal: CType) -> NoReturn:
+        ctx_desc = '\n'.join(f"  {t.cpp_type()} {e};" for t, e in ctx.items())
+        raise UnsatError(f'''
+Failed to synthesize the expression "{goal.cpp_type()} {goal.name}"
+while trying to translate from:
+
+  from_func({', '.join(b.defn() for b in bindings)})
+
+to:
+
+  to_func({', '.join(g.defn() for g in goals)})
+
+When I failed, the following bindings were available in the context:
+
+{ctx_desc}
+
+This probably means there is a missing rule in the rules of tools.codegen.api.translate.
+Check this module for more information.
+''')
+
+    # A shitty backtracking search implementation.  It's shitty because it
+    # doesn't actually do backtracing or search. In particular, if
+    # direct=True, we won't try to do any fancy synthesis, just trivial
+    # conversions (e.g., "T a" is OK for "const T& a").  So all of the
+    # existing rules in this function simply try to solve immediately,
+    # and bail if things don't work out.
+    def solve(goal: CType, *, direct: bool) -> str:
+        def direct_solve(goal: CType) -> str:
+            return solve(goal, direct=True)
+
+        if goal in ctx:
+            # Trivial
+            return ctx[goal]
+
+        # If the goal is a const&, try solving for the value type first
+        if isinstance(goal, ConstRefCType):
+            try:
+                return solve(goal.elem, direct=direct)
+            except UnsatError:
+                pass
+
+        if direct:
+            unsat(goal)
+
+        # For now, all of these rules are mutually exclusive.
+        if goal == OptionalCType(BaseCType("MemoryFormat", "memory_format")):
+            memory_format = direct_solve(
+                OptionalCType(BaseCType("MemoryFormat", SpecialArgName.possibly_redundant_memory_format))
+            )
+            try:
+                options = direct_solve(options_ctype)
+                return f"c10::impl::check_tensor_options_and_extract_memory_format({options}, {memory_format})"
+            except UnsatError:
+                return memory_format
+
+        elif goal == BaseCType("TensorOptions", "options"):
```

Comment:
I haven't verified all the details yet, but the idea looks good me.

The python codegen did similar stuff to some extent (pulling "goal" from available "bindings" in the context): https://github.com/pytorch/pytorch/pull/46244#discussion_r506575407 :p

(note to myself) One natural idea is to apply the translator to python codegen and autograd codegen as well, but the autograd codegen often emits extra initialization statements, and TensorOptions definition there is slightly different (has requires_grad field). Maybe we will need a different translator for the "top-half" of the codegen.

Reply:
Yeah, I'm not sure how possible it is. Need to read the code more carefully to tell!